[ML] Redesign index-based Data Visualizer

qn895 opened the pull request:
## Summary

This PR brings a redesign to the index based data visualizer, where instead of cards for fields, we use table in order to save space and provide more information in a compact way.
![Data Visualizer - Machine Learning - Elastic (5)](https://user-images.githubusercontent.com/43350163/102153091-08b56e80-3e3c-11eb-8b69-1f027ec7eb10.png)

- [x] With the `Show empty fields` enabled:
<img width="1778" alt="Screen Shot 2020-12-14 at 18 40 57" src="https://user-images.githubusercontent.com/43350163/102153121-179c2100-3e3c-11eb-938f-9bfa3e5cdc14.png">

- [x] With `Distributions` preview off:
<img width="1792" alt="Screen Shot 2020-12-13 at 20 03 49" src="https://user-images.githubusercontent.com/43350163/102032430-57043800-3d7e-11eb-841d-7005f33588b5.png">

- [x] With `Distributions` preview on:
<img width="1792" alt="Screen Shot 2020-12-13 at 20 03 12" src="https://user-images.githubusercontent.com/43350163/102032447-63889080-3d7e-11eb-870c-203b99f2cbc4.png">

## To follow up
- Replace the file based data visualizer with the table
- Expand the functional tests to validate the content of the expanded cards

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/master/packages/kbn-i18n/README.md)
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

elasticmachine commented:
Pinging @elastic/ml-ui (:ml)

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/)
* Commit: 0e3db85626ed6bf7f8eed31fae125caff2c8742d
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/execution/node/377/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/","number":"93867","commit":"0e3db85626ed6bf7f8eed31fae125caff2c8742d"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/)
* Commit: 6381ff9fa902757f96031e0dc14d54befe982656
* This comment will update when the build is complete

### Failed CI Steps
* [Execute xpack-kibana-ciGroup3](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/execution/node/629/log)
* [Execute xpack-kibana-ciGroup3](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/execution/node/751/log)

### History
* :broken_heart: [Build #93867](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/) failed 0e3db85626ed6bf7f8eed31fae125caff2c8742d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/","number":"93867","commit":"0e3db85626ed6bf7f8eed31fae125caff2c8742d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/","number":"93871","commit":"6381ff9fa902757f96031e0dc14d54befe982656"}]}
PIPELINE-->

pheyos commented on x-pack/test/functional/apps/ml/anomaly_detection/saved_search_job.ts:
```diff
@@ -77,7 +77,7 @@ export default function ({ getService }: FtrProviderContext) {
         },
         row: {
-          recordCount: '34,416',
+          recordCount: '86,274',
```
Not sure if this change was intended. But it's failing on my local deployment and on CI. I think we should revert this.

qn895 replied:
Reverted here 958b660

pheyos commented on x-pack/test/functional/apps/ml/anomaly_detection/saved_search_job.ts:
```diff
@@ -127,14 +127,14 @@ export default function ({ getService }: FtrProviderContext) {
         },
         row: {
-          recordCount: '34,415',
+          recordCount: '86,274',
```
Not sure if this change was intended. But it's failing on my local deployment and on CI. I think we should revert this.

qn895 replied:
Reverted here 958b660

pheyos commented on x-pack/test/functional/apps/ml/anomaly_detection/saved_search_job.ts:
```diff
@@ -127,14 +127,14 @@ export default function ({ getService }: FtrProviderContext) {
         },
         counts: {
-          processed_record_count: '34,415',
+          processed_record_count: '86,274',
```
Not sure if this change was intended. But it's failing on my local deployment and on CI. I think we should revert this.

qn895 replied:
Reverted here 958b660

pheyos commented on x-pack/test/functional/apps/ml/anomaly_detection/saved_search_job.ts:
```diff
@@ -146,7 +146,7 @@ export default function ({ getService }: FtrProviderContext) {
           earliest_record_timestamp: '2016-02-07 00:00:00',
           latest_record_timestamp: '2016-02-11 23:59:54',
-          input_record_count: '34,415',
+          input_record_count: '86,274',
```
Not sure if this change was intended. But it's failing on my local deployment and on CI. I think we should revert this.

qn895 replied:
Reverted here https://github.com/elastic/kibana/pull/85726/commits/958b660e3b151b77001a6dc7bb460b5ae278a84f

peteharverson commented on x-pack/plugins/ml/public/application/components/multi_select_picker/multi_select_picker.tsx:
```diff
@@ -0,0 +1,131 @@
+  );
+};
+
+export const MultiselectPicker: FC<{
```
Nit - this should be `MultiSelectPicker` if you want to match the file name of `multi_select_picker.tsx`.

qn895 replied:
Renamed here https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb

peteharverson commented on x-pack/plugins/ml/public/application/components/multi_select_picker/multi_select_picker.tsx:
```diff
@@ -0,0 +1,131 @@
+      <div className="euiFilterSelect__noteContent">
+        <EuiIcon type="minusInCircle" />
+        <EuiSpacer size="xs" />
+        <p>No filters found</p>
```
This text needs internationalizing.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/content_types/date_content.tsx:
```diff
@@ -4,64 +4,75 @@
+  const { earliest, latest } = stats;
 
-      <div data-test-subj="mlFieldDataCardEarliest">
+  const summaryTableTitle = i18n.translate('xpack.ml.fieldDataCard.cardNumber.summaryTableTitle', {
```
The keys in this file use `cardNumber`, where it should be `cardDate` I guess?

qn895 replied:
Renamed here [`13fa2c8` (#85726)](https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb)

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/content_types/date_content.tsx:
```diff
@@ -4,64 +4,75 @@
+    {
+      function: 'earliest',
+      display: (
+        <FormattedMessage id="xpack.ml.fieldDataCard.cardNumber.minLabel" defaultMessage="min" />
```
`cardDate` in place of `cardNumber` here?

qn895 replied:
Renamed here https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/content_types/date_content.tsx:
```diff
@@ -4,64 +4,75 @@
-          values={{
-            earliestFormatted: formatDate(earliest, TIME_FORMAT),
-          }}
+          id="xpack.ml.fieldDataCard.cardNumber.medianLabel"
```
`cardDate` in place of `cardNumber` here?

qn895 replied:
Renamed here https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb

pheyos commented on x-pack/test/functional/apps/ml/data_visualizer/index_data_visualizer.ts:
```diff
@@ -144,7 +144,7 @@ export default function ({ getService }: FtrProviderContext) {
     nonMetricFieldsTypeFilter: 'keyword',
     expected: {
-      totalDocCountFormatted: '34,415',
+      totalDocCountFormatted: '86,274',
```
The fact that the tests are passing now with the changed numbers indicates that saved searches are not respected correctly when loading the data (the full data set has 86274 docs, but the saved search that is loaded here, should only show 34415 of them). If saved searches are still supported with the new index data visualizer, I'd suggest to revert this test change and fix the data loading on the app side.

peteharverson replied:
My tests confirm what @pheyos is seeing here, and the saved search doesn't seem to be used in the new page. In the old page, you can see the search in the search bar on opening:

![image](https://user-images.githubusercontent.com/7405507/102091823-3b00a500-3e17-11eb-87e3-ae5c69d49fb5.png)

We should still support loading from a saved search.

qn895 replied:
This saved search issue should now be fixed, I have also reverted the changes here https://github.com/elastic/kibana/pull/85726/commits/958b660e3b151b77001a6dc7bb460b5ae278a84f and here https://github.com/elastic/kibana/pull/85726/commits/c33a88f087611e282de6d01b35432bf5d5a00204

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/examples_list/examples_list.tsx:
```diff
@@ -32,19 +37,17 @@ export const ExamplesList: FC<Props> = ({ examples }) => {
-      </EuiTitle>
+      <ExpandedRowFieldHeader>
+        <FormattedMessage
+          id="xpack.ml.fieldDataCard.cardText.examplesTitle"
```
Nit - maybe don't need `cardText` in this id - `xpack.ml.fieldDataCard.examplesList.title` ?

qn895 replied:
Renamed here https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/search_panel/field_type_filter.tsx:
```diff
@@ -0,0 +1,82 @@
+import { Option, MultiselectPicker } from '../../../../components/multi_select_picker';
+import { FieldTypeIcon } from '../../../../components/field_type_icon';
+
+export const ML_JOB_FIELD_TYPES = {
```
Could the `ML_JOB_FIELD_TYPES` enum in `ml/common/types/field_types.ts` be converted to a `const` and used instead of defining it again here

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/05bb9f39066f6955cb56ab4f09cc4f793dfcdef4

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/page.tsx:
```diff
@@ -165,40 +169,62 @@ export const Page: FC = () => {
+    });
+  };
 
   // TODO - type overallStats and stats
```
Can this `TODO` be removed - looks like you have typed `overallStats` at least in this PR.

qn895 replied:
Removed here https://github.com/elastic/kibana/pull/85726/commits/13fa2c8f86d766e6db81a72638ed3dd8f968cedb

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+    {
+      function: 'min',
+      display: (
+        <FormattedMessage id="xpack.ml.fieldDataCard.cardNumber.minLabel" defaultMessage="min" />
```
The message IDs need editing in this file as they start with `xpack.ml.fieldDataCard.cardNumber` currently.

qn895 replied:
Renamed here https://github.com/elastic/kibana/pull/85726/commits/5eed3fabc69eb50cfa9235e86527c83fd2a14108

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+      function: 'median',
+      display: (
+        <FormattedMessage
+          id="xpack.ml.fieldDataCard.cardNumber.medianLabel"
```
Another id that needs editing.

qn895 replied:
Renamed here 5eed3fa

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+    {
+      function: 'max',
+      display: (
+        <FormattedMessage id="xpack.ml.fieldDataCard.cardNumber.maxLabel" defaultMessage="max" />
```
This `id` needs renaming.

qn895 replied:
Renamed here 5eed3fa

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+    },
+  ];
+
+  const summaryTableTitle = i18n.translate('xpack.ml.fieldDataCard.cardNumber.summaryTableTitle', {
```
This `id` need renaming.

qn895 replied:
Renamed here 5eed3fa

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+            <EuiFlexItem>
+              <ExpandedRowFieldHeader>
+                <FormattedMessage
+                  id="xpack.ml.fieldDataCard.cardNumber.topValuesTitle"
```
This id need renaming.

qn895 replied:
Renamed here 5eed3fa

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+            <EuiFlexItem>
+              <ExpandedRowFieldHeader>
+                <FormattedMessage
+                  id="xpack.ml.fieldDataCard.cardNumber.distributionTitle"
```
This id need renaming.

qn895 replied:
Renamed here 5eed3fa

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+                <EuiFlexItem grow={false}>
+                  <EuiText size="xs">
+                    <FormattedMessage
+                      id="xpack.ml.fieldDataCard.cardNumber.displayingPercentilesLabel"
```
This id need renaming.

qn895 replied:
Renamed here 5eed3fa

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_index_based.ts:
```diff
@@ -37,12 +37,88 @@ export function MachineLearningDataVisualizerIndexBasedProvider(
-        expectedPanelCount,
-        `Expected field panels count to be '${expectedPanelCount}' (got '${allPanels.length}')`
-      );
+    async assertDataVisualizerTotalDocCountHeaderExist() {
```
Nit: Since we're already in the `data_visualizer_index_based` service, I think we don't necessarily need to repeat `DataVisualizer` in the method names.

peteharverson reviewed (commented):
Just some minor code level comments before I give it a test.

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_index_based.ts:
```diff
@@ -37,12 +37,88 @@ export function MachineLearningDataVisualizerIndexBasedProvider(
+        );
+        expect(expectedCount).to.eql(
+          expectedCount,
+          `Expected metric fields count count to be '${expectedCount}' (got '${actualCount}')`
```
```suggestion
          `Expected metric fields count to be '${expectedCount}' (got '${actualCount}')`
```

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_index_based.ts:
```diff
@@ -37,12 +37,88 @@ export function MachineLearningDataVisualizerIndexBasedProvider(
+      });
+    },
+
+    async assertMetricFieldsCount(count: number) {
```
How is this one different from `assertVisibleMetricFieldsCount`?

pheyos replied:
I see that you've updated the method name to clarify the difference between `assertVisibleMetricFieldsCount` and `assertTotalMetricFieldsCount`. But as far as I can see, the implementation is still the same, as both methods are targeting the `mlDataVisualizerVisibleMetricFieldsCount` test subject.

qn895 replied:
You're right I missed that - updated here https://github.com/elastic/kibana/pull/85726/commits/46d06b5dfc7368ad61d09511ef80511ba8c37943

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,162 @@
+    <EuiFlexGroup direction={'row'} data-test-subj={'mlNumberSummaryTable'}>
+      <EuiFlexItem>
+        <ExpandedRowFieldHeader>{summaryTableTitle}</ExpandedRowFieldHeader>
+        <EuiBasicTable<SummaryTableItem>
```
Is it possible to hide the empty table header row here? Otherwise you get a thin white row below the `SUMMARY` title:

![image](https://user-images.githubusercontent.com/7405507/102091247-7babee80-3e16-11eb-8739-d1e606b09eba.png)

qn895 replied:
Talked to the EUI team and this is not possible. However, since we are changing the background color to match the background anyway, this should be less noticeable.

<img width="1466" alt="Screen Shot 2020-12-14 at 13 23 15" src="https://user-images.githubusercontent.com/43350163/102126060-308edd00-3e10-11eb-9f91-707f457cd939.png">

peteharverson commented:
For the data for each row, do you only query for the fields visible on the current page? It looks like data for all pages might still be loaded. For indexes with large amounts of fields, just querying for the fields on the current page would be a good optimization (not needed for this PR, but worth adding to the list of follow-ups if it doesn't do this already).

![image](https://user-images.githubusercontent.com/7405507/102093652-4fde3800-3e19-11eb-8048-14321e57dd45.png)

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_count_panel/field_count_panel.tsx:
```diff
@@ -0,0 +1,127 @@
+              <h5>
+                <FormattedMessage
+                  id="xpack.ml.dataVisualizer.searchPanel.metricsLabel"
+                  defaultMessage="Metrics"
```
I know we used this label in the old page, but I wonder if something like `number fields` would make more sense, corresponding to the text in the field type dropdown?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/bd93f5afb82a40117f16b37ec2337b6907715224
![Uploading Screen Shot 2020-12-14 at 10.27.20.png…]()

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_count_panel/field_count_panel.tsx:
```diff
@@ -0,0 +1,127 @@
+              <h5>
+                <FormattedMessage
+                  id="xpack.ml.dataVisualizer.searchPanel.fieldsLabel"
+                  defaultMessage="Fields"
```
And similarly, not sure about this label now. `Other fields`?

peteharverson commented:
If I sort on the `Documents` column, and then hit page 2, I get this error:

![image](https://user-images.githubusercontent.com/7405507/102096200-57531080-3e1c-11eb-98d5-2b61d3817ecb.png)

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/top_values/top_values.tsx:
```diff
@@ -43,32 +43,32 @@ export const TopValues: FC<Props> = ({ stats, fieldFormat, barColor }) => {
+      {Array.isArray(topValues) &&
+        topValues.map((value: any) => (
+          <EuiFlexGroup gutterSize="xs" alignItems="center" key={value.key}>
+            <EuiFlexItem grow={false} style={{ width: 100 }} className="eui-textTruncate">
```
With the full width content in the expanded row, you can probably increase the width of the label, to give more space for text like URIs:

![image](https://user-images.githubusercontent.com/7405507/102096972-37701c80-3e1d-11eb-8209-9deeda797bb9.png)

mdefazio commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/content_types/date_content.tsx:
```diff
@@ -4,64 +4,75 @@
 
-      <div data-test-subj="mlFieldDataCardEarliest">
+  const summaryTableTitle = i18n.translate('xpack.ml.fieldDataCard.cardNumber.summaryTableTitle', {
+    defaultMessage: 'summary',
```
There seems to be a few spots where these aren't Title Case.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/20ad0dbab26023ed3b5b9c1635af8423002d908d

mdefazio commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/expanded_row_field_header/expanded_row_field_header.tsx:
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+
+export const ExpandedRowFieldHeader = ({ children }: { children: React.ReactNode }) => (
+  <EuiText size="xs" color={'subdued'} className={'mlFieldDataCard__valuesTitle'}>
+    {children}
+  </EuiText>
```
I think we can remove this class altogether with some slight tweaks to the design.  Discover does not color these type of headings, or center them. So I think we can simply do:

```
<EuiTitle size="xxxs">
    <h4>{children}</h4>
</EuiTitle>
```

qn895 replied:
I think we will keep it like this for now with the class to remain consistent with the file-based data visualizer.

mdefazio reviewed (changes requested):
A few edits around the css.  Ideally we can remove some of those classes that are just doing some slight style changes from our defaults (likely my fault in the design/handoff).

Also, I wasn't sure exactly where it needs to happen, but it doesn't look like we have the prop `isExpandedRow` being added to the expanded row [(docs)](https://eui.elastic.co/#/tabular-content/tables#expanding-rows). With this, I think we can remove the border between the row and the expanded content so they better connect visually.  And I think by adding the tinted background, we lose some contrast on the charts.  So ideally we could keep everything on white.  
![image](https://user-images.githubusercontent.com/3756330/102101758-08e04a80-3df9-11eb-9ff6-3c705ef6a1c3.png)

Happy to discuss this further—I get there might be some concerns with being able to scan through the list and see what the expanded content is compared to a collapsed row.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/)
* Commit: 0a04a37939e8e8fd550408b1b2fae388ae057170
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/execution/node/371/log)

### History
* :broken_heart: [Build #93871](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/) failed 6381ff9fa902757f96031e0dc14d54befe982656
* :broken_heart: [Build #93867](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/) failed 0e3db85626ed6bf7f8eed31fae125caff2c8742d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/","number":"93867","commit":"0e3db85626ed6bf7f8eed31fae125caff2c8742d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/","number":"93871","commit":"6381ff9fa902757f96031e0dc14d54befe982656"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/","number":"94074","commit":"0a04a37939e8e8fd550408b1b2fae388ae057170"}]}
PIPELINE-->

peteharverson commented:
The alignment of the top values chart looks a bit odd here, where it only contains a few categories? Can it be aligned to the top?

![image](https://user-images.githubusercontent.com/7405507/102115047-e539f600-3e32-11eb-8152-62a773a38f66.png)

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/_expanded_row.scss:
```diff
@@ -0,0 +1,16 @@
+    font-weight: bold;
+  }
+  .mlFieldDataCard__codeContent {
+    font-family: $euiCodeFontFamily;
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/0c34e8a5e3eaca6ac6bacc336870461c19f09b96

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/_expanded_row.scss:
```diff
@@ -0,0 +1,16 @@
+.mlDataVisualizerFieldExpandedRow {
+  padding-left: 50px;
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/0c34e8a5e3eaca6ac6bacc336870461c19f09b96

qn895 commented:
> The alignment of the top values chart looks a bit odd here, where it only contains a few categories? Can it be aligned to the top?

This is now updated here https://github.com/elastic/kibana/pull/85726/commits/609cef83f6c7972ae8cf507bdc49a74876745992

<img width="1466" alt="Screen Shot 2020-12-14 at 13 23 15" src="https://user-images.githubusercontent.com/43350163/102126532-cdea1100-3e10-11eb-9236-16800af41017.png">

> If I sort on the Documents column, and then hit page 2, I get this error:
Fixed here https://github.com/elastic/kibana/pull/85726/commits/37ac2828359249d66365c062628d6616ba06170e. This issue is caused by us trying to serialize the sorting's state, which in this case for EUI is giving back a jsx component instead of a string field. I removed the icons in the column header to make it work correctly.

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/)
* Commit: 0c34e8a5e3eaca6ac6bacc336870461c19f09b96
* This comment will update when the build is complete

### Failed CI Steps
* [X-Pack Jest Unit Tests](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/execution/node/329/log)

### History
* :broken_heart: [Build #94074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/) failed 0a04a37939e8e8fd550408b1b2fae388ae057170
* :broken_heart: [Build #93871](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/) failed 6381ff9fa902757f96031e0dc14d54befe982656
* :broken_heart: [Build #93867](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/) failed 0e3db85626ed6bf7f8eed31fae125caff2c8742d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/","number":"93867","commit":"0e3db85626ed6bf7f8eed31fae125caff2c8742d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/","number":"93871","commit":"6381ff9fa902757f96031e0dc14d54befe982656"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/","number":"94074","commit":"0a04a37939e8e8fd550408b1b2fae388ae057170"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/","number":"94155","commit":"0c34e8a5e3eaca6ac6bacc336870461c19f09b96"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/)
* Commit: eea52b84cece2f74e7cc232bd621dfdfb9eae3bb
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/execution/node/367/log)

### History
* :broken_heart: [Build #94155](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/) failed 0c34e8a5e3eaca6ac6bacc336870461c19f09b96
* :broken_heart: [Build #94074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/) failed 0a04a37939e8e8fd550408b1b2fae388ae057170
* :broken_heart: [Build #93871](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/) failed 6381ff9fa902757f96031e0dc14d54befe982656
* :broken_heart: [Build #93867](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/) failed 0e3db85626ed6bf7f8eed31fae125caff2c8742d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/","number":"93867","commit":"0e3db85626ed6bf7f8eed31fae125caff2c8742d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/","number":"93871","commit":"6381ff9fa902757f96031e0dc14d54befe982656"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/","number":"94074","commit":"0a04a37939e8e8fd550408b1b2fae388ae057170"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/","number":"94155","commit":"0c34e8a5e3eaca6ac6bacc336870461c19f09b96"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/","number":"94271","commit":"eea52b84cece2f74e7cc232bd621dfdfb9eae3bb"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/)
* Commit: 36497eb5310f26674bbde6f3bab8d35a49eff05b

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/b8ab3fdafc7562b1cfc85b354971a93c7a218f9e) | [after](https://github.com/elastic/kibana/commit/36497eb5310f26674bbde6f3bab8d35a49eff05b) | diff |
| --- | --- | --- | --- |
| `ml` | 1593 | 1606 | +13 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/b8ab3fdafc7562b1cfc85b354971a93c7a218f9e) | [after](https://github.com/elastic/kibana/commit/36497eb5310f26674bbde6f3bab8d35a49eff05b) | diff |
| --- | --- | --- | --- |
| `ml` | 5.3MB | 5.3MB | +33.1KB |

#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/b8ab3fdafc7562b1cfc85b354971a93c7a218f9e) | [after](https://github.com/elastic/kibana/commit/36497eb5310f26674bbde6f3bab8d35a49eff05b) | diff |
| --- | --- | --- | --- |
| `default` | 47138 | 47903 | +765 |

### History
* :broken_heart: [Build #94271](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/) failed eea52b84cece2f74e7cc232bd621dfdfb9eae3bb
* :broken_heart: [Build #94155](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/) failed 0c34e8a5e3eaca6ac6bacc336870461c19f09b96
* :broken_heart: [Build #94074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/) failed 0a04a37939e8e8fd550408b1b2fae388ae057170
* :broken_heart: [Build #93871](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/) failed 6381ff9fa902757f96031e0dc14d54befe982656
* :broken_heart: [Build #93867](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/) failed 0e3db85626ed6bf7f8eed31fae125caff2c8742d

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93867/","number":"93867","commit":"0e3db85626ed6bf7f8eed31fae125caff2c8742d"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/","number":"93871","commit":"6381ff9fa902757f96031e0dc14d54befe982656"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/","number":"94074","commit":"0a04a37939e8e8fd550408b1b2fae388ae057170"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/","number":"94155","commit":"0c34e8a5e3eaca6ac6bacc336870461c19f09b96"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/","number":"94271","commit":"eea52b84cece2f74e7cc232bd621dfdfb9eae3bb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/","number":"94286","commit":"36497eb5310f26674bbde6f3bab8d35a49eff05b"}]}
PIPELINE-->

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/common/field_vis_config.ts:
```diff
@@ -4,18 +4,65 @@
+  sampleCount?: number;
+  trueCount?: number;
+  falseCount?: number;
+  earliest?: any;
+  latest?: any;
```
are these two a number type?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/common/field_vis_config.ts:
```diff
@@ -4,18 +4,65 @@
+  earliest?: any;
+  latest?: any;
+  documentCounts?: {
+    buckets?: any;
```
`any` 👀

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_count_panel/field_count_panel.tsx:
```diff
@@ -0,0 +1,128 @@
+    <EuiFlexGroup
+      alignItems="center"
+      gutterSize="xs"
+      style={{ marginLeft: 5 }}
```
not sure if we ever used this value for margins

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/search_panel/field_name_filter.tsx:
```diff
@@ -0,0 +1,62 @@
+
+interface Props {
+  overallStats: OverallStats;
+  setVisibleFieldNames(q: any): void;
```
is it an array of strings?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/search_panel/field_name_filter.tsx:
```diff
@@ -0,0 +1,62 @@
+  visibleFieldNames,
+  showEmptyFields,
+}) => {
+  const [items, setItems] = useState<Option[]>([]);
+
+  useEffect(() => {
+    const options: Option[] = [];
+    if (overallStats) {
+      Object.keys(overallStats).forEach((key) => {
+        const fieldsGroup = overallStats[key as keyof OverallStats];
+        if (Array.isArray(fieldsGroup) && fieldsGroup.length > 0) {
+          fieldsGroup.forEach((field) => {
+            if (
+              (field.existsInDocs === true || showEmptyFields === true) &&
+              field.fieldName !== undefined
+            ) {
+              options.push({ value: field.fieldName });
+            }
+          });
+        }
+      });
+    }
+    setItems(options);
+  }, [overallStats]);
```
```suggestion
 const items: Option[] = useMemo(() => {
     const options: Option[] = [];
    if (overallStats) {
      Object.keys(overallStats).forEach((key) => {
        const fieldsGroup = overallStats[key as keyof OverallStats];
        if (Array.isArray(fieldsGroup) && fieldsGroup.length > 0) {
          fieldsGroup.forEach((field) => {
            if (
              (field.existsInDocs === true || showEmptyFields === true) &&
              field.fieldName !== undefined
            ) {
              options.push({ value: field.fieldName });
            }
          });
        }
      });
    }
    return options;
 }, [overallStats])
```

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/search_panel/search_panel.tsx:
```diff
@@ -33,54 +24,50 @@ import {
   samplerShardSize: number;
   setSamplerShardSize(s: number): void;
-  totalCount: number;
+  overallStats: any;
+  indexedFieldTypes: MlJobFieldType[];
+  setVisibleFieldTypes(q: any): void;
+  visibleFieldTypes: string[];
+  setVisibleFieldNames(q: any): void;
```
please avoid using `any`

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,173 @@
+      <VerticalSpacer />
+      {stats && (
+        <EuiFlexItem>
+          <EuiFlexGroup direction={'column'} data-test-subj={'mlTopValues'}>
```
question: do we need a flex layout here?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/a021855fbe9e0d6d08dbfd2f6b847f25c3f12a80

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,173 @@
+
+      {distribution && (
+        <EuiFlexItem>
+          <EuiFlexGroup direction={'column'} data-test-subj={'mlMetricDistribution'}>
```
question: do we need a flex layout here?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/a021855fbe9e0d6d08dbfd2f6b847f25c3f12a80

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_expanded_row/number_content.tsx:
```diff
@@ -0,0 +1,173 @@
+          tableCaption={summaryTableTitle}
+        />
+      </EuiFlexItem>
+      <VerticalSpacer />
```
Do you actually need this component? 🤔 Have you tried setting a gutter size instead?

mdefazio replied:
Agree... `gutterSize='xl'` should work

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/stats_datagrid.tsx:
```diff
@@ -0,0 +1,251 @@
+  updatePageState,
+}: DataVisualizerDataGrid) => {
+  const [expandedRowItemIds, setExpandedRowItemIds] = useState<string[]>([]);
+  const [expandAll, toggleExpandAll] = useState<boolean | undefined>();
```
can it be just `boolean`?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

darnautov commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/stats_datagrid.tsx:
```diff
@@ -0,0 +1,251 @@
+    const index = expandedRowItemIds.indexOf(item.fieldName);
+    if (index !== -1) {
+      expandedRowItemIds.splice(index, 1);
+      setExpandedRowItemIds([...expandedRowItemIds]);
```
you have `setExpandedRowItemIds` at the end of the function, so it called twice. I guess this one should be removed. 
```suggestion
```

qn895 replied:
Updated here 4ea6d17

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_table.ts:
```diff
@@ -0,0 +1,164 @@
+      await this.expandRowDetails(fieldName, fieldType);
+
+      await testSubjects.existOrFail(
+        `mlDataVisualizerFieldExpandedRow ${fieldName} ${fieldType} > mlFieldDataCardTopValues`
```
I'd suggest to remove `Card` from this data test subject as it's not longer content of a card.

pheyos commented on x-pack/test/functional/services/ml/data_visualizer_table.ts:
```diff
@@ -0,0 +1,164 @@
+      await this.expandRowDetails(fieldName, fieldType);
+
+      await testSubjects.existOrFail(
+        `mlDataVisualizerFieldExpandedRow ${fieldName} ${fieldType} > mlFieldDataCardExamplesList`
```
I'd suggest to remove `Card` from this data test subject as it's not longer content of a card.

pheyos commented on x-pack/test/functional/apps/ml/data_visualizer/index_data_visualizer.ts:
```diff
@@ -326,105 +345,60 @@ export default function ({ getService }: FtrProviderContext) {
-        );
-        await ml.dataVisualizerIndexBased.clearFieldsPanelSearchInput(fieldTypes);
-        await ml.dataVisualizerIndexBased.assertSampleSizeInputExists();
-        await ml.dataVisualizerIndexBased.setSampleSizeInputValue(
```
Taking this line as an example for a few things.
I'm a bit worried about losing test coverage here as the following user actions are not tested anymore:
- Filter table with the search bar
- Filter table with the field type input
- Change the sample size

I think it would be good to merge without losing test coverage, but not sure if this is in the scope of this PR or if this should be done in a follow-up. @peteharverson could you please assess?

Also, there are a couple `ml.dataVisualizerIndexBased` service methods that are not needed anymore (or need to be adjusted to cover the user actions mentioned above):
- assertFieldsPanelForTypesExist
- assertCardExists
- assertCardContentsExists
- assertNonMetricCardContents
- assertDocumentCountCardContents
- assertNumberCardContents
- assertDateCardContents
- assertKeywordCardContents
- assertTextCardContents
- assertFieldDocCountExists
- assertFieldDocCountContents
- assertFieldCardinalityExists
- assertNumberStatsContents
- assertTopValuesContents
- assertFieldsPanelCardCount
- assertFieldsPanelSearchInputValue
- clearFieldsPanelSearchInput
- filterFieldsPanelWithSearchString
- assertFieldsPanelTypeInputExists
- assertFieldsPanelTypeInputValue
- setFieldsPanelTypeInputValue
- setSampleSizeInputValue

peteharverson replied:
We definitely don't want to lose test coverage, but I'm happy to merge this PR without those three tests @pheyos highlighted, and to add them back in a subsequent PR.

peteharverson commented on x-pack/plugins/ml/public/application/components/data_grid/use_column_chart.tsx:
```diff
@@ -120,11 +120,11 @@ export const isUnsupportedChartData = (arg: any): arg is UnsupportedChartData =>
 
 type LegendText = string | JSX.Element;
-const getLegendText = (chartData: ChartData): LegendText => {
+export const getLegendText = (chartData: ChartData): LegendText => {
```
The legend always says `top 20 of... ` but we are showing a max of 10 columns here:

![image](https://user-images.githubusercontent.com/7405507/102225498-330c3800-3edf-11eb-8fee-f391ecbcdf06.png)

qn895 replied:
Fixed here https://github.com/elastic/kibana/pull/85726/commits/286a76b18e6f8c7563503c965e654a8f4eecdd46

peteharverson commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/components/field_data_row/number_content_preview.tsx:
```diff
@@ -0,0 +1,73 @@
+        {legendText && (
+          <>
+            <EuiFlexGroup direction={'row'} data-test-subj={`${dataTestSubj}-legend`}>
+              <EuiFlexItem className={'mlDataGridChart__legend'}>{legendText.min}</EuiFlexItem>
```
Is it possible to format the `min` and `max` using the formatter for that field in the index pattern, as we do in the chart in the expanded row:

![image](https://user-images.githubusercontent.com/7405507/102227260-43bdad80-3ee1-11eb-9039-70d6c7a642b7.png)

peteharverson reviewed (approved):
Tested latest edits and overall LGTM.

Found another couple of minor issues that would be good to fix.

walterra commented:
<img width="25%" src="https://user-images.githubusercontent.com/230104/102235866-17129180-3ef3-11eb-9270-afe722358640.png" />

Some suggestions:

- The blue chart's axis labels are missing the padding the other chart has
- Should we align the styling so both charts have the same thin white padding between bars?
- Is there a need to use different colors? I'd opt to use the same color like we do in the data grid. If different colors, IMHO it should match the icon colors in the `type` column.

mdefazio commented:
Overall layout is looking good.  Just a few tiny items:

- Add some space between the 'Distribution' column heading and the icon button to show distribution charts
![image](https://user-images.githubusercontent.com/3756330/102237733-f6324800-3ec2-11eb-89a2-9f1cf8279d84.png)
![image](https://user-images.githubusercontent.com/3756330/102237790-064a2780-3ec3-11eb-9dc7-4a8dcfc0a92a.png)
With 4px between
![image](https://user-images.githubusercontent.com/3756330/102237914-2c6fc780-3ec3-11eb-818b-f36350f459ff.png)

- The top chart, search/filter bar and field counters feel cramped—I would suggest perhaps a large spacer between these.
![image](https://user-images.githubusercontent.com/3756330/102238306-98523000-3ec3-11eb-9d37-5f0ae575631b.png)

- I think the distribution charts a bit big and could use some extra padding on the top / bottom so they aren't so close to the edges of the rows
  - Some of these have their x-axis labels touching the chart
![image](https://user-images.githubusercontent.com/3756330/102237245-760be280-3ec2-11eb-89de-207f180293a5.png)
  - Can we get the axis font size to be the same as the axis font size on the main document distribution chart? 
![image](https://user-images.githubusercontent.com/3756330/102237466-b2d7d980-3ec2-11eb-949e-1e6e87184c0c.png)

walterra commented on x-pack/plugins/ml/public/application/data_frame_analytics/pages/analytics_management/components/analytics_list/use_table_settings.ts:
```diff
@@ -7,6 +7,7 @@
 import { Direction, EuiBasicTableProps, Pagination, PropertySort } from '@elastic/eui';
 import { useCallback, useMemo } from 'react';
 import { ListingPageUrlState } from '../../../../../../../common/types/common';
+import { DataVisualizerIndexBasedAppState } from '../../../../../../../common/types/ml_url_generator';
```
Since `useTableSettings()` is now used in other places than just from within the analytics management page, should we move the file to a more general place?

walterra commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/content_types/date_content.tsx:
```diff
@@ -4,64 +4,77 @@
-            latestFormatted: formatDate(latest, TIME_FORMAT),
-          }}
+  return (
+    <EuiFlexGroup direction={'row'}>
```
Can you explain why we need a flex layout with a single item here?

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/0dcdf2efbdbfdbd6a0259f38cc37bac7942cfae7

walterra commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/top_values/_top_values.scss:
```diff
@@ -0,0 +1,11 @@
+.mlTopValuesLabelContainer {
+  padding-right: $euiSizeM;
+
+  &.small {
+    width: 50px !important;
+  }
+
+  &.large {
+    width: 200px !important;
+  }
+}
```
We should use a different structure, with this we end up with non-prefixed very generic css classes in the DOM like `small/large`. Consider using the BEM-style described on the bottom of this page (Sass best practices) https://elastic.github.io/eui/#/guidelines/sass.

You should end up creating classes that look like `mlTopValuesLabelContainer--small`. In the react code you will need to use both the base class and this modifier.

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/0dcdf2efbdbfdbd6a0259f38cc37bac7942cfae7

walterra commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/top_values/top_values.tsx:
```diff
@@ -43,32 +45,39 @@ export const TopValues: FC<Props> = ({ stats, fieldFormat, barColor }) => {
+              className={classNames(
+                'eui-textTruncate',
+                'mlTopValuesLabelContainer',
+                compressed === true ? 'small' : 'large'
```
With the SASS changes suggested in the other comment this should become e.g. `mlTopValuesLabelContainer--${compressed === true ? 'small' : 'large'}`

qn895 replied:
Updated here https://github.com/elastic/kibana/pull/85726/commits/0dcdf2efbdbfdbd6a0259f38cc37bac7942cfae7

qn895 commented on x-pack/plugins/ml/common/types/datavisualizer.ts:
```diff
@@ -0,0 +1,13 @@
+  aggregatableExistsFields: any[];
+  aggregatableNotExistsFields: any[];
+  nonAggregatableExistsFields: any[];
+  nonAggregatableNotExistsFields: any[];
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/a021855fbe9e0d6d08dbfd2f6b847f25c3f12a80

qn895 commented on x-pack/plugins/ml/public/application/components/data_grid/column_chart.tsx:
```diff
@@ -68,7 +68,7 @@ export const ColumnChart: FC<Props> = ({ chartData, columnType, dataTestSubj })
         {legendText}
       </div>
-      <div data-test-subj={`${dataTestSubj}-id`}>{columnType.id}</div>
+      {hideLabel !== true && <div data-test-subj={`${dataTestSubj}-id`}>{columnType.id}</div>}
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

qn895 commented on x-pack/plugins/ml/public/application/components/multi_select_picker/multi_select_picker.tsx:
```diff
@@ -0,0 +1,137 @@
+  checked?: 'on' | 'off';
+  disabled?: boolean;
+  type?: string;
+  onChange?: (items: any[]) => void;
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/4ea6d174728033f29d81392da47db1f3b409d4a5

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/)
* Commit: 3305f1a136b6f0b02648b8adc9ae03837343425f
* This comment will update when the build is complete

### Failed CI Steps
* [Check i18n](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/execution/node/394/log)

### History
* :green_heart: [Build #94286](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/) succeeded 36497eb5310f26674bbde6f3bab8d35a49eff05b
* :broken_heart: [Build #94271](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/) failed eea52b84cece2f74e7cc232bd621dfdfb9eae3bb
* :broken_heart: [Build #94155](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/) failed 0c34e8a5e3eaca6ac6bacc336870461c19f09b96
* :broken_heart: [Build #94074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/) failed 0a04a37939e8e8fd550408b1b2fae388ae057170
* :broken_heart: [Build #93871](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/) failed 6381ff9fa902757f96031e0dc14d54befe982656

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/93871/","number":"93871","commit":"6381ff9fa902757f96031e0dc14d54befe982656"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/","number":"94074","commit":"0a04a37939e8e8fd550408b1b2fae388ae057170"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/","number":"94155","commit":"0c34e8a5e3eaca6ac6bacc336870461c19f09b96"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/","number":"94271","commit":"eea52b84cece2f74e7cc232bd621dfdfb9eae3bb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/","number":"94286","commit":"36497eb5310f26674bbde6f3bab8d35a49eff05b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/","number":"94569","commit":"3305f1a136b6f0b02648b8adc9ae03837343425f"}]}
PIPELINE-->

qn895 commented:
> * Add some space between the 'Distribution' column heading and the icon button to show distribution charts
> * The top chart, search/filter bar and field counters feel cramped—I would suggest perhaps a large spacer between these.
> * I think the distribution charts a bit big and could use some extra padding on the top / bottom so they aren't so close to the edges of the rows
>   
>   * Some of these have their x-axis labels touching the chart
>   * Can we get the axis font size to be the same as the axis font size on the main document distribution chart?


@mdefazio @walterra I have updated the colors of the chart here, with additional paddings in between here  https://github.com/elastic/kibana/pull/85726/commits/59e11f97b7bd8699af8dd4e80cc70f9a4cccc2d7

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/_index.scss:
```diff
@@ -1,2 +1,3 @@
 @import 'file_based/index';
 @import 'index_based/index';
+@import "stats_datagrid/index";
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/0dcdf2efbdbfdbd6a0259f38cc37bac7942cfae7

darnautov reviewed (approved):
LGTM

mdefazio reviewed (approved):
There were a few small things (also mentioned on Slack) that can be cleaned up. Fine if this is done in a follow-up PR.

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/_field_data_card.scss:
```diff
@@ -57,6 +57,7 @@
 
   .mlFieldDataCard__codeContent {
+    @include euiCodeFont;
     font-family: $euiCodeFontFamily;
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/c16d60d1c8737037d72e492663fe60fcde730495

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/index_based/components/field_data_card/top_values/_top_values.scss:
```diff
@@ -0,0 +1,13 @@
+
+  &.mlTopValuesLabelContainer--large {
+    width: 200px !important;
+  }
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/c16d60d1c8737037d72e492663fe60fcde730495

qn895 commented on x-pack/plugins/ml/public/application/datavisualizer/stats_datagrid/_index.scss:
```diff
@@ -0,0 +1,38 @@
+  }
+  .mlFieldDataCard__codeContent {
+    @include euiCodeFont;
+    font-family: $euiCodeFontFamily;
```
Updated here https://github.com/elastic/kibana/pull/85726/commits/c16d60d1c8737037d72e492663fe60fcde730495

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94590/)
* Commit: b2abf178079896e097e7911895ab704fa644faaf

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/507ae79ffe040af9d4f05d01830e1f42e0b34d65) | [after](https://github.com/elastic/kibana/commit/b2abf178079896e097e7911895ab704fa644faaf) | diff |
| --- | --- | --- | --- |
| `ml` | 1593 | 1604 | +11 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/507ae79ffe040af9d4f05d01830e1f42e0b34d65) | [after](https://github.com/elastic/kibana/commit/b2abf178079896e097e7911895ab704fa644faaf) | diff |
| --- | --- | --- | --- |
| `ml` | 5.3MB | 5.3MB | +31.4KB |

#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/507ae79ffe040af9d4f05d01830e1f42e0b34d65) | [after](https://github.com/elastic/kibana/commit/b2abf178079896e097e7911895ab704fa644faaf) | diff |
| --- | --- | --- | --- |
| `default` | 47241 | 48005 | +764 |

### History
* :broken_heart: [Build #94569](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/) failed 3305f1a136b6f0b02648b8adc9ae03837343425f
* :green_heart: [Build #94286](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/) succeeded 36497eb5310f26674bbde6f3bab8d35a49eff05b
* :broken_heart: [Build #94271](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/) failed eea52b84cece2f74e7cc232bd621dfdfb9eae3bb
* :broken_heart: [Build #94155](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/) failed 0c34e8a5e3eaca6ac6bacc336870461c19f09b96
* :broken_heart: [Build #94074](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/) failed 0a04a37939e8e8fd550408b1b2fae388ae057170

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94074/","number":"94074","commit":"0a04a37939e8e8fd550408b1b2fae388ae057170"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/","number":"94155","commit":"0c34e8a5e3eaca6ac6bacc336870461c19f09b96"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/","number":"94271","commit":"eea52b84cece2f74e7cc232bd621dfdfb9eae3bb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/","number":"94286","commit":"36497eb5310f26674bbde6f3bab8d35a49eff05b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/","number":"94569","commit":"3305f1a136b6f0b02648b8adc9ae03837343425f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94590/","number":"94590","commit":"b2abf178079896e097e7911895ab704fa644faaf"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94650/)
* Commit: c16d60d1c8737037d72e492663fe60fcde730495

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/507ae79ffe040af9d4f05d01830e1f42e0b34d65) | [after](https://github.com/elastic/kibana/commit/c16d60d1c8737037d72e492663fe60fcde730495) | diff |
| --- | --- | --- | --- |
| `ml` | 1593 | 1604 | +11 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/507ae79ffe040af9d4f05d01830e1f42e0b34d65) | [after](https://github.com/elastic/kibana/commit/c16d60d1c8737037d72e492663fe60fcde730495) | diff |
| --- | --- | --- | --- |
| `ml` | 5.3MB | 5.3MB | +31.1KB |

#### Distributable file count
| id | [before](https://github.com/elastic/kibana/commit/507ae79ffe040af9d4f05d01830e1f42e0b34d65) | [after](https://github.com/elastic/kibana/commit/c16d60d1c8737037d72e492663fe60fcde730495) | diff |
| --- | --- | --- | --- |
| `default` | 47241 | 48005 | +764 |

### History
* :green_heart: [Build #94590](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94590/) succeeded b2abf178079896e097e7911895ab704fa644faaf
* :broken_heart: [Build #94569](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/) failed 3305f1a136b6f0b02648b8adc9ae03837343425f
* :green_heart: [Build #94286](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/) succeeded 36497eb5310f26674bbde6f3bab8d35a49eff05b
* :broken_heart: [Build #94271](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/) failed eea52b84cece2f74e7cc232bd621dfdfb9eae3bb
* :broken_heart: [Build #94155](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/) failed 0c34e8a5e3eaca6ac6bacc336870461c19f09b96

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94155/","number":"94155","commit":"0c34e8a5e3eaca6ac6bacc336870461c19f09b96"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94271/","number":"94271","commit":"eea52b84cece2f74e7cc232bd621dfdfb9eae3bb"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94286/","number":"94286","commit":"36497eb5310f26674bbde6f3bab8d35a49eff05b"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94569/","number":"94569","commit":"3305f1a136b6f0b02648b8adc9ae03837343425f"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94590/","number":"94590","commit":"b2abf178079896e097e7911895ab704fa644faaf"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/94650/","number":"94650","commit":"c16d60d1c8737037d72e492663fe60fcde730495"}]}
PIPELINE-->
